### PR TITLE
[None][feat] Support multi-gpu running for nemotron-v3-nano and super

### PR DESCRIPTION
**Pull request by Wanli-Jiang**
# Features
* Verified 1/2/4 GPUs with TP, TEP for nano-v3 bf16/fp8/nvfp4 ckpts.
* Verified 1/2/4/8 GPUs with TP, TEP for super-v3 bf16 ckpts.
* Support cudagraph with multi-GPUs.
* Support multi-stream for MoE shared_experts and routed_experts.

# TODOs
* Support attention_dp for multi-gpugs. (will raise error or hanging for now, a draft PR is https://github.com/NVIDIA/TensorRT-LLM/pull/10220).

## Summary by CodeRabbit

* **Bug Fixes**
  * Resolved weight tensor loading and alignment issues for quantized models to ensure computational accuracy.
  * Enhanced weight scale handling in NVFP4 quantization mode.
  * Corrected weight processing in mixture-of-experts layers.

* **Refactor**
  * Optimized internal weight handling efficiency and consistency across quantization methods.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/modules/linear.py [hunk 1]
```diff
@@ -1144,24 +1144,31 @@ def post_load_weights(self, module: Linear):
...
+                    module.scaling_vector_size)
+                weight_scale_unswizzle_pad = F.pad(
+                    weight_scale_unswizzle,
+                    (0, scale_pad_col, 0, scale_pad_row),
+                    mode='constant',
+                    value=0)
+                module.weight_scale = Parameter(
+                    torch.ops.trtllm.block_scale_interleave(
+                        weight_scale_unswizzle_pad),
+                    requires_grad=False)
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/modules/linear.py]: The change in this `linear.py` file is from https://github.com/NVIDIA/TensorRT-LLM/pull/10148, which is a fix for nvfp4 weight_scale in Mamba layers.

File: tensorrt_llm/_torch/models/checkpoints/hf/weight_mapper.py [hunk 2]
```diff
@@ -69,12 +69,17 @@ def _duplicate_kv_weights(self, module: nn.Module, new_name: str,
                 num_kv_heads = kv_shape * 2 // self._head_dim
             else:
                 num_kv_heads = kv_shape // self._head_dim
+
+            duplicated_keys = ["weight", "bias"]
+            if module.quant_config.quant_mode.has_nvfp4():
+                duplicated_keys.append("weight_scale")
```
Comment (nv-guomingz) [inline tensorrt_llm/_torch/models/checkpoints/hf/weight_mapper.py]: Is it a nemotron-only change?

Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/checkpoints/hf/weight_mapper.py]: no, I think all nvfp4 models need it if  `num_kv_head` of the attention layer in an nvfp4 models is smaller than tp_size.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 3]
```diff
@@ -701,6 +710,43 @@ def load_quant_scales(self, module: torch.nn.Module, weights: Dict):
         module.fc2_dequant.data.copy_(tmp_w2_weight_scale * max_fc2_input_scale)
         module.fc31_input_dequant.data.copy_(max_fc31_input_scale)
 
+    def post_load_weights(self, module):
+        if getattr(module, "moe_backend", None) != "CUTLASS":
+            return
+
+        super().post_load_weights(module)
+        """
+        Pad weight and weight_scale tensors to meet torch trtllm NVFP4 GEMM alignment requirements.
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py]: If this docstring is for `_maybe_padding_weights`, it should be put inside `_maybe_padding_weights`.

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py [hunk 4]
```diff
@@ -110,19 +117,17 @@ def preprocess_weights(self, weights: dict) -> dict:
                         elif "weight_scale" in key:
                             # NVFP4 case.
                             if weights[name].shape:
-                                new_weights[w3_key] = weights[
-                                    name][:weights[name].shape[0] // 2]
-                                new_weights[w1_key] = weights[name][
-                                    weights[name].shape[0] // 2:]
+                                # w3 weight (gate_proj) scale should be empty for Nemotron-H MoE model.
+                                new_weights[w3_key] = weights[name][:0]
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: Can we use `torch.empty(...)` instead of `weights[name][:0]`?

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py [hunk 5]
```diff
@@ -110,19 +117,17 @@ def preprocess_weights(self, weights: dict) -> dict:
...
                             else:
                                 new_weights[w3_key] = weights[name]
                                 new_weights[w1_key] = weights[name]
                         else:
-                            new_weights[w3_key] = weights[name][:weights[name].
-                                                                shape[0] // 2]
-                            new_weights[w1_key] = weights[name][weights[name].
-                                                                shape[0] // 2:]
+                            # w3 weight (gate_proj) should be empty for Nemotron-H MoE model.
+                            new_weights[w3_key] = weights[name][:0]
```
Comment (yuxianq) [inline tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: Can we use `torch.empty(...)` instead of `weights[name][:0]`?

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 6]
```diff
@@ -475,12 +475,21 @@ def load_expert_w3_w1_weight(self,
             TensorParallelMode.COLUMN,
             device=device) if w3_weight is not None else None
 
-        dst_w3_weight, dst_w1_weight = dst_w3_w1_weight.chunk(2, dim=0)
+        src_w3_size_shard = w3_weight_shard.shape[
+            0] if w3_weight_shard is not None else 0
+        src_w1_size_shard = w1_weight_shard.shape[
+            0] if w1_weight_shard is not None else 0
         if w1_weight is not None:
+            dst_w1_weight = dst_w3_w1_weight.narrow(dim=0,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py]: Do you mean `dst_w1_weight` is an empty tensor if `w1_weight_shard` is None? But I think `w1_weight_shard` cannot be None if `w1_weight` is not None, right?

Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: There is a `partial_load` flag in main, which can cause w1_weight as None.

Here we want to be compatible with two cases, w1_weight/w3_weight are None or w1_weight/w3_weight are with zero shape in output dim.  Such two cases will not in-place copy src_weights to dst_weights.

**Review by yuxianq — APPROVED** [review]

**Review by nv-guomingz — APPROVED** [review]

File: tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py (hunk 4 above)
Comment (Wanli-Jiang) [reply tensorrt_llm/_torch/models/checkpoints/hf/nemotron_h_weight_mapper.py]: done

**Comment by Wanli-Jiang** [comment]
/bot run --stage-list "DGX_H100-2_GPUs-PyTorch-Others-1" --disable-fail-fast

**Comment by Wanli-Jiang** [comment]
/bot run --stage-list "DGX_H100-2_GPUs-PyTorch-Others-1" --disable-fail-fast

**Comment by Wanli-Jiang** [comment]
/bot skip --comment "all tests are passed"
